daemon/policy.go

```diff
@@ -302,6 +304,79 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 	return newRev, nil
 }
 
+type policyAddQueueOptions struct {
+	rules   policyAPI.Rules
+	opts    *AddOptions
+	context *policyQueueContext
+	reFetch bool
+}
+
+type policyQueueContext struct {
+	policyRevision uint64
+	err            error
+	done           chan bool
+	totalTime      spanstat.SpanStat
+	queueTime      spanstat.SpanStat
+}
+
+var policyAddQueueRequests = make(chan policyAddQueueOptions, 100)
+
+// policyAddToQueue adds the request to policyAdd in the queue and return a
+// context that will be updated when the done channel is closed.
+func (d *Daemon) policyAddToQueue(rules policyAPI.Rules, opts *AddOptions, reFetch bool) *policyQueueContext {
+	result := &policyQueueContext{
+		policyRevision: 0,
+		err:            nil,
+		done:           make(chan bool),
+	}
+	result.totalTime.Start()
+	result.queueTime.Start()
+
+	policyAddQueueRequests <- policyAddQueueOptions{
+		rules:   rules,
+		opts:    opts,
+		reFetch: reFetch,
+		context: result,
+	}
+
+	return result
+}
+
+// PolicyQueueWorker is a queue worker that taking care of PolicyAdd for
+// DNSProxy and Kubernetes watcher to avoid multiples races and lock all policy
+// repo.
+// In case that the channel event has a refetch set to true, the rule will be
+// retrieved from policyRepo to make sure that the newest version is installed
+// correctly and old rules do not overwritte the new ones.
+func (d *Daemon) PolicyQueueWorker(ctx context.Context) {
+	log.Info("Initialize Policy Queue Worker")
+	for {
+		select {
+		case policyRequest, _ := <-policyAddQueueRequests:
+			policyRequest.context.queueTime.End(true)
+			var rules []*policyAPI.Rule
+			rules = policyRequest.rules
+			if policyRequest.reFetch {
+				newRules := policyAPI.Rules{}
+				d.policy.Mutex.Lock()
+				for _, sourceRule := range policyRequest.rules {
+					repoRules := d.policy.SearchRLocked(sourceRule.Labels)
```

Comment:
This is an absolute no go for performance. This is O(n*m) with n = new rules, m = existing rules. We need to find a better way.

Reply:
Hi,

I was thinking on this but it's not a trivial thing, for me the best thing is to
get something like this:

```
GetMultipleRulesFromLabels(lbls ...Labels) {
    allLabels := map[string]lbls {}

    for _,lbl := range lbls {
        allLabels[lbl.SHA()] == lbl
    }


	result := api.Rules{}

	for _, r := range p.rules {
        _, exists := allLabels[r.SHA()]
	    if exists {
			result = append(result, &r.Rule)
		}
	}
	return result
}
```

But this SHA can be super hard to get, because need to do a loop over all the
labels, so it's will not scale at all. For me the best way to fix that is to
have the SHA in the LabelArray type, and each one that one is added or deleted
that can be updated.

Something like this diff:
https://gist.github.com/eloycoto/bd5bab5c2a6e33fbb3317e2cfd2d3a6d

Any insight here?

Best
Eloy

Reply:
@eloycoto I've done preliminary work to add a SHA into the `LabelArray` here https://github.com/cilium/cilium/commit/d33176c1054b9ef185a12eb6225fda48f0856a73 if you're interested

Reply:
@ianvernon I'm ok with your changes, but this will not be backported no? This need to be in 1.4.2 release.

Reply:
@eloycoto oof, yeah, I didn't think about that...